File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
def solution(gems):
    n = len(gems)
    gem_type: int = len(set(gems))
    length = float('inf')
    answer = []
    bag = dict()
    left = right = 0
    while right < n:
        if gems[right] in bag:
            bag[gems[right]] += 1
        else:
            bag[gems[right]] = 1
        while len(bag) == gem_type:
            if right - left < length:
                length = right - left
                answer = [left+1, right+1]
            bag[gems[left]] -= 1
            if bag[gems[left]] == 0:
                del bag[gems[left]]
            left += 1
        right += 1
            
    return answer

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;

class Solution {
    public int[] solution(String[] gems) {
        Set<String> gemType = new HashSet<>(Arrays.asList(gems));
        int total = gemType.size();
        Map<String, Integer> counter = new HashMap<>();
        int l = 0, r = 0;
        int minLen = gems.length;
        int answerL = 0, answerR = gems.length-1;
        while (r < gems.length) {
            String gem = gems[r];
            counter.put(gem, counter.getOrDefault(gem, 0) + 1);
            while (counter.size() == total && l <= r) {
                if (r - l < minLen) {
                    minLen = r - l;
                    answerL = l;
                    answerR = r;
                }
                String lGem = gems[l];
                counter.put(lGem, counter.get(lGem) - 1);
                if (counter.get(lGem) == 0) counter.remove(lGem);
                l++;
            }
            r++;
        }
        
        return new int[]{answerL+1, answerR+1};
    }
}